### PR TITLE
storage/mysql: Fix transient errors being reported as definite when validating system config

### PR DESCRIPTION
**Pull request by rjobanp**
### Motivation

  * This PR fixes a previously unreported bug.

I encountered a spurious bug when working on another issue that can happen if the mysql server is restarted or experiences a network issue when we validate system settings. We were blindly reporting all errors as 'Definite Errors' in these instances, even though transient network errors can be retried.

I captured the source statuses when this happened, and you can see the error state `mysql server configuration error: Input/output error: Input/output error` is the message from the `DefiniteError::ServerConfigurationError` value wrapped around a transient network error.

```
materialize=> SELECT * FROM mz_internal.mz_source_statuses;
 id  |             name             |   type    |   last_status_change_at    |  status  |                                                          error 
                                                          |                                                                   details                    
                                                
-----+------------------------------+-----------+----------------------------+----------+----------------------------------------------------------------
----------------------------------------------------------+----------------------------------------------------------------------------------------------
------------------------------------------------
 u14 | mysql_table7                 | subsource | 2024-02-28 20:34:09.285+00 | starting |                                                                
                                                          | 
 u19 | mysql_table5                 | subsource | 2024-02-28 20:34:09.325+00 | starting |                                                                
                                                          | 
 u16 | mysql_table7_source          | mysql     | 2024-02-28 20:34:09.285+00 | starting |                                                                
                                                          | 
 u21 | mysql_table5_source          | mysql     | 2024-02-28 20:34:09.325+00 | starting |                                                                
                                                          | 
 u5  | mysql_table1_source_progress | progress  |                            | running  |                                                                
                                                          | 
 u15 | mysql_table7_source_progress | progress  |                            | running  |                                                                
                                                          | 
 u20 | mysql_table5_source_progress | progress  |                            | running  |                                                                
                                                          | 
 u4  | mysql_table1                 | subsource | 2024-02-28 20:33:39.144+00 | ceased   | mysql server configuration error: Input/output error: Input/out
put error: Connection reset by peer (os error 104)        | 
 u6  | mysql_table1_source          | mysql     | 2024-02-28 20:33:39.144+00 | stalled  | mysql: mysql server configuration error: Input/output error: In
put/output error: Connection reset by peer (os error 104) | {"namespaced":{"mysql":"mysql server configuration error: Input/output error: Input/output er
ror: Connection reset by peer (os error 104)"}}
(9 rows)

```

### Tips for reviewer

The code should be self explanatory -- we now will match explicitly on the `MySqlError::InvalidSystemSetting` error type before returning a definite error

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/storage/src/source/mysql/replication.rs [hunk 1]
```diff
@@ -417,10 +417,14 @@ async fn raw_stream<'a>(
...
-    if let Err(err) = validate_mysql_repl_settings(&mut conn).await {
-        return Ok(Err(DefiniteError::ServerConfigurationError(
-            err.to_string(),
-        )));
+    match validate_mysql_repl_settings(&mut conn).await {
+        Err(err) if matches!(err, MySqlError::InvalidSystemSetting { .. }) => {
+            return Ok(Err(DefiniteError::ServerConfigurationError(
+                err.to_string(),
+            )));
+        }
```
Comment (guswynn) [inline src/storage/src/source/mysql/replication.rs:425]:
```suggestion
        Err(err @ MySqlError::InvalidSystemSetting { .. }) => {
            return Ok(Err(DefiniteError::ServerConfigurationError(
                err.to_string(),
            )));
        }
```

lil rust pattern spicy sauce with `@` bindings

File: src/storage/src/source/mysql/snapshot.rs [hunk 2]
```diff
@@ -291,17 +291,21 @@ pub(crate) fn render<G: Scope<Timestamp = GtidPartition>>(
...
-                        &all_outputs,
-                        &mut raw_handle,
-                        data_cap_set,
-                        &mut definite_error_handle,
-                        definite_error_cap_set,
-                    )
-                    .await);
-                }
+                match validate_mysql_repl_settings(&mut conn).await {
+                    Err(err) if matches!(err, MySqlError::InvalidSystemSetting { .. }) => {
```
Comment (guswynn) [inline src/storage/src/source/mysql/snapshot.rs:295]: same here with `@`

**Review by guswynn — APPROVED** [review]
Do we want to invert this logic? default to returning a definite error but have known transient cases? I suspect not, but I want to make sure we have made a conscious decision to default to transient here

**Comment by rjobanp** [comment]
> default to returning a definite error but have known transient cases? 

Well the only time we want to return the definite-error is when the network calls succeed and we very the configuration is incorrect.

Reversing the logic will be a lot more verbose since the `MySqlError` enum can hold several other error-types (via thiserror) that all belong under the transient-error category (I/O failures, other network blips, etc). If you feel strongly about it I can do so.

File: src/storage/src/source/mysql/replication.rs (hunk 1 above)
Comment (rjobanp) [reply src/storage/src/source/mysql/replication.rs:425]: oh nice!

**Comment by guswynn** [comment]
@rjobanp sounds good!

**Pull request merged**